File: 6.Software-Design-Patterns/example/Creational/Builder/Builder.java
public class User {
  private final String name;
  private final int age;
  private final int money;
  private final boolean hasJob;
  private final boolean isEmployee;

  private User(Builder builder) {
    this.name = builder.name;
    this.age = builder.age;
    this.money = builder.money;
    this.hasJob = builder.hasJob;
    this.isEmployee = builder.isEmployee;
  }

  public static class Builder {
    private String name;
    private int age;
    private int money; // Optional
    private boolean hasJob = false;
    private boolean isEmployee = false;

    public Builder setName(String name) {
      this.name = name;
      return this;
    }

    public Builder setAge(int age) {
      this.age = age;
      return this;
    }

    public Builder setMoney(int money) {
      this.money = money;
      return this;
    }

    public Builder makeJob() {
      this.hasJob = true;
      return this;
    }

    public Builder makeEmployee() {
      this.hasJob = true;
      this.isEmployee = true;
      return this;
    }

    public User build() {
      if (name == null) {
        throw new IllegalArgumentException("Name is a required field");
      }
      return new User(this);
    }
  }

  @Override
  public String toString() {
    return "User{" +
        "name='" + name + '\'' +
        ", age=" + age +
        ", money=" + money +
        ", hasJob=" + hasJob +
        ", isEmployee=" + isEmployee +
        '}';
  }
}

// Usage Example
public class Main {
  public static void main(String[] args) {
    User user = new User.Builder()
        .setName("Mahdi")
        .setAge(25)
        .setMoney(200)
        .makeEmployee()
        .build();

    System.out.println(user);
  }
}

File: 2.Intermediate Level/6.Software-Design-Patterns/example/Creational/Builder/Builder.js
class UserBuilderJs{
    #name;#age;#hasJob;#money;#isEmployee
    constructor() {
        this.#name = ''
        this.#age = 0
        this.#hasJob = false
        this.#money = 0
        this.#isEmployee = false
    }
    setName(name){
        this.#name = name;
        return this;
      }
    
      setAge(age){
        this.#age = age;
        return this;
      }
    
      setMoney(money){
        this.#money = money;
        return this;
      }
    
      makeJob(){
        this.#hasJob = true;
        return this;
      }
    
      makeEmployee(){
        this.#hasJob = true;
        this.#isEmployee = true;
        return this;
      }
      build() {
        return {
          name: this.#name,
          age: this.#age,
          money: this.#money,
          isEmployee: this.#isEmployee,
          hasJob: this.#hasJob,
        };
      }
}
class PersonJs{
    constructor(builder) {
        this.name = builder.name
        this.age = builder.age
        this.money = builder.money
        this.hasJob = builder.hasJob
        this.isEmployee = builder.isEmployee
    }
}
const person2 = new PersonJs(new UserBuilderJs().setName('mahdi').setAge(20).setMoney(200).build())

console.log(person2);